File: script/create_missing_partitions.pl
#!/usr/bin/env perl
#
# Create missing partitions for for partitioned tables
#

use strict;
use warnings;

BEGIN
{
    use FindBin;
    use lib "$FindBin::Bin/../lib";
}

use Modern::Perl "2015";
use MediaWords::CommonLibs;
use MediaWords::DB;

use Sys::RunAlone;

use Readonly;

sub main
{
    binmode( STDOUT, 'utf8' );
    binmode( STDERR, 'utf8' );

    # Wait for an hour between attempts to create new partitions
    Readonly my $DELAY_BETWEEN_ATTEMPTS => 60 * 60;

    INFO "Starting to create missing partitions...";
    while ( 1 )
    {
        INFO "Creating missing partitions...";

        my $db = MediaWords::DB::connect_to_db;
        $db->query( 'SELECT create_missing_partitions()' );
        $db->disconnect;

        INFO "Created missing partitions, sleeping for $DELAY_BETWEEN_ATTEMPTS seconds.";
        sleep( $DELAY_BETWEEN_ATTEMPTS );
    }
}

main();

# Required by Sys::RunAlone
__END__

File: apps/create-missing-partitions/bin/create_missing_partitions.py
#!/usr/bin/env python3
#
# Create missing partitions for for partitioned tables
#

import time

from mediawords.db import connect_to_db
from mediawords.util.log import create_logger

log = create_logger(__name__)


def create_missing_partitions():
    """Call PostgreSQL function which creates missing table partitions (if any)."""

    # Wait for an hour between attempts to create new partitions
    delay_between_attempts = 60 * 60

    log.info("Starting to create missing partitions...")
    while True:
        log.info("Creating missing partitions...")

        db = connect_to_db()
        db.query('SELECT create_missing_partitions()')
        db.disconnect()

        log.info("Created missing partitions, sleeping for %d seconds." % delay_between_attempts)
        time.sleep(delay_between_attempts)


if __name__ == '__main__':
    create_missing_partitions()
